File: src/org/foonugget/kdict/data/Dictionary.java

package org.foonugget.kdict.data;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.google.inject.Inject;

import java.util.ArrayList;
import java.util.List;

public class Dictionary {

    @SuppressWarnings("unused")
    private final static String TAG = Dictionary.class.getSimpleName();

    private SQLiteDatabase mDB;

    @Inject
    public Dictionary(DatabaseOpenHelper helper) {
        mDB = helper.openDatabase();
    }

    public List<WordMatch> wordSearch(String word) {
        word = word.trim();
        addSearchWordToHistory(word);

        ArrayList<WordMatch> matches = new ArrayList<WordMatch>();
        matches.addAll(findKoreanByEnglishExact(word));
        matches.addAll(findEnglishByKoreanExact(word));

        return matches;
    }

    public List<WordMatch> wordSearchEach(String word) {
        word = word.trim();
        addSearchWordToHistory(word);

        ArrayList<WordMatch> matches = new ArrayList<WordMatch>();

        for (int i = 0; i < word.length(); i++) {
            String ch = word.substring(i, i + 1);
            matches.addAll(findKoreanByEnglishExact(ch));
            matches.addAll(findEnglishByKoreanExact(ch));
        }

        return matches;
    }

    public List<WordMatch> wordSearchContains(String word) {
        word = word.trim();
        addSearchWordToHistory(word);

        ArrayList<WordMatch> matches = new ArrayList<WordMatch>();
        matches.addAll(findKoreanByEnglishFuzzy(word));
        matches.addAll(findEnglishByKoreanFuzzy(word));

        return matches;
    }

    public List<WordMatch> findKoreanByEnglishFuzzy(String word) {
        final String SQL = "SELECT word, definition FROM english_korean WHERE word LIKE '%"
                + escapeString(word) + "%'";
        return getWordMatches(SQL);
    }

    public List<WordMatch> findEnglishByKoreanFuzzy(String word) {
        final String SQL = "SELECT word, definition FROM korean_english WHERE word LIKE '%"
                + escapeString(word) + "%'";
        return getWordMatches(SQL);
    }

    public List<WordMatch> findKoreanByEnglishExact(String word) {
        final String SQL = "SELECT word, definition FROM english_korean WHERE word = '"
                + escapeString(word) + "'";
        return getWordMatches(SQL);
    }

    public List<WordMatch> findEnglishByKoreanExact(String word) {
        final String SQL = "SELECT word, definition FROM korean_english WHERE word = '"
                + escapeString(word) + "'";
        return getWordMatches(SQL);
    }

    protected String escapeString(String string) {
        return string.replaceAll("'", "''");
    }
    
    private void addSearchWordToHistory(String word) {
        final String SQL = "INSERT INTO search_history VALUES ((SELECT MAX(id)+1 FROM search_history), '" + escapeString(word) + "')";
        mDB.execSQL(SQL);
    }

    private List<WordMatch> getWordMatches(final String SQL) {
        List<WordMatch> matches = new ArrayList<WordMatch>(0);

        Cursor cursor = null;
        try {
            cursor = mDB.rawQuery(SQL, null);
            if (cursor.moveToFirst() == false) {
                cursor.close();
                return matches;
            }

            while (cursor.isAfterLast() == false) {
                String word = cursor.getString(0);
                String translation = cursor.getString(1);
                WordMatch wordMatch = new WordMatch(word, translation);
                matches.add(wordMatch);
                cursor.moveToNext();
            }
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        return matches;
    }

}

File: app/src/main/java/org/foonugget/kdict/data/Dictionary.kt
package org.foonugget.kdict.data

import android.database.sqlite.SQLiteDatabase

class Dictionary(
    private val mDictDB: SQLiteDatabase
) {
    fun wordSearch(word: String): List<WordMatch> {
        val trimmed = word.trim()
        if (trimmed.isEmpty()) return emptyList()
        val matches = mutableListOf<WordMatch>()
        matches.addAll(findKoreanByEnglishExact(trimmed))
        matches.addAll(findEnglishByKoreanExact(trimmed))
        return matches
    }

    fun wordSearchEach(word: String): List<WordMatch> {
        val trimmed = word.trim()
        if (trimmed.isEmpty()) return emptyList()
        val matches = mutableListOf<WordMatch>()
        for (i in 0 until trimmed.length) {
            val ch = trimmed.substring(i, i + 1)
            matches.addAll(findKoreanByEnglishExact(ch))
            matches.addAll(findEnglishByKoreanExact(ch))
        }
        return matches
    }

    fun wordSearchContains(word: String): List<WordMatch> {
        val trimmed = word.trim()
        if (trimmed.isEmpty()) return emptyList()
        val matches = mutableListOf<WordMatch>()
        matches.addAll(findKoreanByEnglishFuzzy(trimmed))
        matches.addAll(findEnglishByKoreanFuzzy(trimmed))
        return matches
    }

    fun findKoreanByEnglishFuzzy(word: String): List<WordMatch> {
        return getWordMatches(
            "SELECT word, definition FROM english_korean WHERE word LIKE ?",
            arrayOf("%$word%")
        )
    }

    fun findEnglishByKoreanFuzzy(word: String): List<WordMatch> {
        return getWordMatches(
            "SELECT word, definition FROM korean_english WHERE word LIKE ?",
            arrayOf("%$word%")
        )
    }

    fun findKoreanByEnglishExact(word: String): List<WordMatch> {
        return getWordMatches(
            "SELECT word, definition FROM english_korean WHERE word = ?",
            arrayOf(word)
        )
    }

    fun findEnglishByKoreanExact(word: String): List<WordMatch> {
        return getWordMatches(
            "SELECT word, definition FROM korean_english WHERE word = ?",
            arrayOf(word)
        )
    }

    private fun getWordMatches(sql: String, selectionArgs: Array<String>): List<WordMatch> {
        val matches = mutableListOf<WordMatch>()
        mDictDB.rawQuery(sql, selectionArgs).use { cursor ->
            while (cursor.moveToNext()) {
                val word = cursor.getString(0)
                val translation = cursor.getString(1)
                matches.add(WordMatch(word, translation))
            }
        }
        return matches
    }
}
